tests/test_silicon_platform_additional.py

```diff
@@ -0,0 +1,358 @@
+        from chipflow_lib.platforms.silicon import SiliconPlatformPort
+
+        port_obj1 = Port(type="input", pins=["1"], port_name="test_port1",
+                        direction=io.Direction.Input, options={})
+        port_obj2 = Port(type="output", pins=["2"], port_name="test_port2",
+                        direction=io.Direction.Output, options={})
+
+        platform._ports = {
+            "test_port1": SiliconPlatformPort("comp", "test_port1", port_obj1),
+            "test_port2": SiliconPlatformPort("comp", "test_port2", port_obj2),
+        }
+
+        # Create a module with a clock domain
+        m = Module()
+        m.domains.sync = ClockDomain()
+
+        # The core thing we want to test is setting the pinlock to our mock
+        if hasattr(platform, "pinlock"):
+            del platform.pinlock
+        self.assertFalse(hasattr(platform, "pinlock"))
+
+        # Call the method we want to test
+        # This should now just set the pinlock attribute
+        # and not try to create additional ports because we mocked an empty pinlock
+        platform.instantiate_ports(m)
+
+        # Check that ports are accessible
+        self.assertEqual(len(platform._ports), 2)
+        self.assertIn("test_port1", platform._ports)
+        self.assertIn("test_port2", platform._ports)
+
+        # Check that pinlock was set
+        self.assertEqual(platform.pinlock, mock_pinlock)
+
+    def test_instantiate_ports_missing_clock(self):
+        """Test instantiate_ports method with missing clock directly"""
+        # Import here to avoid issues during test collection
+        from chipflow_lib.platforms.silicon import SiliconPlatform, load_pinlock, ChipFlowError
+        from amaranth import Module
+
+        # Create a config with missing clock configuration
+        # This deliberately causes an error to test error handling
+        config_copy = self.config.copy()
+        config_copy["chipflow"] = config_copy.get("chipflow", {}).copy()
+        config_copy["chipflow"]["clocks"] = {"default": "non_existent_clock"}
+        config_copy["chipflow"]["resets"] = {}
+
+        # Create platform with our modified config
+        platform = SiliconPlatform(config_copy)
+
+        # Make sure pinlock is not already set
+        if hasattr(platform, "pinlock"):
+            del platform.pinlock
+
+        # Create a Module
+        m = Module()
+
+        # Create a custom TestPinlock with an empty clocks dict
+        class TestPinlock:
+            def __init__(self):
+                self.port_map = {}
+                self.package = mock.MagicMock()
+                self.package.clocks = {}
+                self.package.resets = {}
+
+        # Patch the load_pinlock function directly
+        original_load_pinlock = load_pinlock
+        try:
+            # Replace with our custom implementation
+            load_pinlock.__globals__['load_pinlock'] = lambda: TestPinlock()
```

Comment:
hm, why can't this use a mock approach like everywhere else?

Reply:
mocking Pydantic turns out to be a bit tricksy.. Not sure this is entirely the right approach though.